Support copy_ for Sparse Compressed tensors.

pearu opened the pull request:
It also fixes bugs in validating sparse block compressed indices and generating sparse block tensors.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77605
* #77530

pearu committed:
Support copy_ for Sparse Compressed tensors.

[ghstack-poisoned]

pearu committed:
Update on "Support copy_ for Sparse Compressed tensors."

It also fixes bugs in validating sparse block compressed indices and generating sparse block tensors.




[ghstack-poisoned]

pearu committed:
Update on "Support copy_ for Sparse Compressed tensors."

It also fixes bugs in validating sparse block compressed indices and generating sparse block tensors.




[ghstack-poisoned]

pearu committed:
Update on "Support copy_ for Sparse Compressed tensors."

It also fixes bugs in validating sparse block compressed indices and generating sparse block tensors.




[ghstack-poisoned]

pearu committed:
Update on "Support copy_ for Sparse Compressed tensors."

It also fixes bugs in validating sparse block compressed indices and generating sparse block tensors.




[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -513,18 +513,39 @@ const Tensor& resize_sparse_csr_(
+      "torch.copy_: only sparse compressed tensors with the same number of specified elements are supported.");
+  auto self_compressed_dims = self.size(compressedDimension(self.layout(), self.sizes()));
+  auto src_compressed_dims = src.size(compressedDimension(src.layout(), src.sizes()));
+  TORCH_CHECK(self_compressed_dims == src_compressed_dims,
```
I'd state the dimension index as well, since we don't really have dedicated APIs to Python users for compressed and plain indices etc. I'd say something like "Expected shape X and shape Y to match along dimension Z for given layouts A and B".

pearu replied:
Fixed.

pearu committed:
Update on "Support copy_ for Sparse Compressed tensors."

It also fixes bugs in validating sparse block compressed indices and generating sparse block tensors.




[ghstack-poisoned]

pearu commented:
@pytorchbot merge on green

pearu commented:
This merge may be premature as I may have misinterpreted the meaning of `@pytorchbot merge on green` - "merge when all CI is green" - but apparently the merge was triggered before the CI finished. Any advice?

^ @cpuhrsch

janeyx99 commented:
@pearu Noted! Sorry for the confusion and a change is in progress to a true all green option: https://github.com/pytorch/pytorch/pull/77660#pullrequestreview-977696201
